ietfdata/tools/affiliations.py

```
@@ -0,0 +1,152 @@
+# Mappings generated:
+# 1. raw affiliation -> normalised affiliation mappings
+# 2. email_domain -> normalised affiliations mappings
+# 3. identity -> start year-month, end year-month, affiliation mapping
+
+# Affiliation Entry Class
+class AffiliationEntry:
+    start_date  : datetime.date
+    end_date    : Optional[datetime.date]
+    names : list[str]
+    
+    def __init__(self, affiliation_str:str, start_date:datetime.date, end_date:Optional[datetime.date],names_list:Optional[list[str]]):
+        if names_list is not None:
+            self.names=copy.deepcopy(names_list)
+        else:
+            self.names = cleanup_affiliation(affiliation_str)
+        self.start_date = start_date
+        self.end_date = end_date
+    
+    def set_end_date(self, new_end_date:datetime.date):
+        self.end_date = new_end_date
+    
+    def set_affiliation_names(self, affiliation:str):
+        self.namse=cleanup_affiliation(affiliation)
+    
+    def match_names(self, names:list[str]):
+        for name in names:
+            if name not in self.names:
+                return False
+        return True
+    
+    def __str__(self):
+        return_str = '{"names":['
+        for name in self.names:
+            return_str+=f'"{name}",'
+        return_str =  return_str.rstrip(',')
+        return_str += "],"
+        return_str += f'"start_date":"{self.start_date}","end_date":"{self.end_date}"}}'
+        return return_str
+
+# Affiliation mapping class
+class AffiliationMap:
+    identifiers : list[str]
+    affiliations: list[AffiliationEntry]
+    
+    def __init__(self,identifiers:list[str],affiliations:list[AffiliationEntry]):
+        self.identifiers = copy.deepcopy(identifiers)
+        self.affiliations = copy.deepcopy(affiliations)
+    
+    def add_identifier(self, identifier:str):
+        self.identifiers.append(identifier)
+    
+    def add_affiliation(self, aff_entry:AffiliationEntry):
```

This assumes affiliations are unique and have been normalised. I wonder if it would be more robust if it normalised affiliations as they were added and silently dropped duplicates?